src/refract/JSONSchemaVisitor.cc

```diff
@@ -216,14 +231,67 @@ namespace refract
         }
     }
 
+    ArrayElement* JSONSchemaVisitor::arrayFromProps(std::vector<MemberElement*>& props)
+    {
+        ArrayElement *a = new ArrayElement;
+        a->renderType(IElement::rCompact);
+        for(std::vector<MemberElement *>::const_iterator i = props.begin();
+            i != props.end();
+            ++i) {
+            ObjectElement *ob = new ObjectElement;
+            ob->renderType(IElement::rCompact);
+            StringElement *str = TypeQueryVisitor::as<StringElement>((*i)->value.first);
+            JSONSchemaVisitor renderer(pDefs, fixed);
+            renderer.visit(*(*i)->value.second);
+            addDefinition(str->value, renderer.getOwnership());
+            StringElement *s = new StringElement;
+            s->renderType(IElement::rCompact);
+            s->set("#/definitions/" + str->value);
+            ob->addMember("$ref", s, IElement::rCompact);
+            a->push_back(ob);
+        }
+
+        return a;
+    }
+
+    void JSONSchemaVisitor::addVariableProps(std::vector<MemberElement*>& props,
+                                             ObjectElement *o)
+    {
+        if (o->empty()) {
+            if (props.size() == 1) {
+                StringElement *str = TypeQueryVisitor::as<StringElement>(props.front()->value.first);
+                JSONSchemaVisitor renderer(pDefs, fixed);
+                renderer.visit(*props.front()->value.second);
+                addDefinition(str->value, renderer.getOwnership());
+                StringElement *s = new StringElement;
```

Comment:
Use setting constructor.
Maybe extract into function like `createReferenceToDefinition()` and use it in `arrayFromProps()` too